lib/src/connection.dart

```diff
@@ -162,40 +165,54 @@ class PostgreSQLConnection implements PostgreSQLExecutionContext {
       throw new PostgreSQLException(
           "Attempting to reopen a closed connection. Create a new instance instead.");
     }
+    try {
+      _hasConnectedPreviously = true;
+      _socket = await Socket
+          .connect(host, port)
+          .timeout(new Duration(seconds: timeoutInSeconds), onTimeout: _timeout);
+
+      _framer = new MessageFramer();
+      if (useSSL) {
+        _socket = await _upgradeSocketToSSL(_socket, timeout: timeoutInSeconds);
+      }
 
-    _hasConnectedPreviously = true;
-    _socket = await Socket
-        .connect(host, port)
-        .timeout(new Duration(seconds: timeoutInSeconds), onTimeout: _timeout);
+      var connectionComplete = new Completer();
 
-    _framer = new MessageFramer();
-    if (useSSL) {
-      _socket = await _upgradeSocketToSSL(_socket, timeout: timeoutInSeconds);
-    }
+      _socket.listen(_readData,
+          onError: _handleSocketError, onDone: _handleSocketClosed);
 
-    var connectionComplete = new Completer();
+      _transitionToState(
+          new _PostgreSQLConnectionStateSocketConnected(connectionComplete));
 
-    _socket.listen(_readData,
-        onError: _handleSocketError, onDone: _handleSocketClosed);
+      return await connectionComplete.future
+          .timeout(new Duration(seconds: timeoutInSeconds), onTimeout: _timeout);
+    }
+    on Exception catch(_) {
+      _close();
+      rethrow;
+    }
+  }
 
-    _transitionToState(
-        new _PostgreSQLConnectionStateSocketConnected(connectionComplete));
+  Future _ensureOpen() async {
+    if (isClosed) {
+      throw new PostgreSQLException(
+          "Attempting to execute query, but connection is closed.");
+    }
 
-    return connectionComplete.future
-        .timeout(new Duration(seconds: timeoutInSeconds), onTimeout: _timeout);
+    if(!_hasConnectedPreviously) {
+      await open();
```

Comment:
I don't think the connection should automatically open for any reason, it must be an explicit `open`.